Require first & last name on project requests

helbashandy opened the pull request:
## Description

* Adds a permissions.py file as an initial step to centeralize
  permissions across the app.

* Adds first and last name check to the test_func on requests to create or join a project.

* Creates a generic wrapper decorator to be used around test_func to
  allow gradual progressive refactor of test_func towards a  more
  centralized and modular permissions management solution.

Fixes #605 

## Type of change

 **** Please delete options that are not relevant. ****

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

**** Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration ****

1. Last name was removed from a user account and saved.
2. Clicked on project and attempted to create project - Saw the error message. 
3. Clicked on project and attempted to join a project - Saw the error message. 


## PR Self Evaluation
strikethrough things that don’t make sense for your PR

- [x] My code follows the agreed upon best practices
- [x] I have commented my code, particularly in hard-to-understand areas
~- [ ] I have made corresponding changes to the documentation (if needed)~
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in the appropriate modules
- [x] I have performed a self-review of my own code

adeng27 committed:
Method for renewing ICAs

adeng27 committed:
disallow non ICA projects

adeng27 committed:
Find PI

adeng27 committed:
Ensure requester is a part of the project. Stop requests for out of date allocation periods.

adeng27 committed:
Remove todo

adeng27 committed:
changes based off pr comments

adeng27 committed:
Initial unit testing

adeng27 committed:
Basic unit testing for key assumptions

matthew-li committed:
Refactor logic for fetching the AllocationPeriods pertinent to a particular allowance type into a method

matthew-li committed:
Refactor logic for setting PI eligibility in renewal request into method

matthew-li committed:
Refactor logic for computing SUs to allocate to new project request/renewal request into method

matthew-li committed:
Generalize projects command > renew to enable renewals of non-pooling FCAs and PCAs; add constraints

matthew-li committed:
Add comments describing additional test cases for projects command > renew

adeng27 committed:
flesh out testing for project renewals

matthew-li committed:
Disallow renewals of limited-per-PI allowances until checks added; add/fix tests

matthew-li committed:
Decorate renewal tests as being BRC tests

matthew-li committed:
Remove entry from .gitignore

helbashandy commented:
Hi @matthew-li - Please note that I've updated the implementation by replacing the use of `permission_classes` that I had initially in mind with the `permission_required` decorator.

This change was made for two key reasons:

- `permission_classes` are specific to DRF and are not compatible with regular Django views.
- I saw this as an opportunity to explore the centralized approach to permissions management. By using the `permission_required` decorator, we can seamlessly integrate it with our existing test_func methods while also passing any common permissions from utils/permissions.py. This approach will facilitate a gradual refactor of our test_func logic across the app progressively.

matthew-li commented on coldfront/core/utils/permissions.py:
```diff
@@ -0,0 +1,32 @@
+    :return:
+    """
+    if request.user.first_name == '' or request.user.last_name == '':
+        messages.error(request, 'You must set your first and last name on your account before you can make requests.')
```
Link directly to the update form within the message?

matthew-li commented on coldfront/core/utils/permissions.py:
```diff
@@ -0,0 +1,32 @@
+    :param permission_check: function that returns True if the user has the permission, False otherwise
+    :return:
+    """
+    def decorator(test_func):
```
Take multiple as input with `*permission_checks` and then loop over them?

So it'd be:

```python
@permissions_required(perm1, perm2, ..., permK)  # Note the plural
def test_func(self):
    pass
```

As opposed to:

```python
@permission_required(permK)
.
.
.
@permission_required(perm2)
@permission_required(perm1)
def test_func(self):
    pass
```

Not especially opinionated on which is cleaner.

helbashandy replied:
I like your suggestion better actually.

helbashandy committed:
Require first & last name on project requests

 * Adds a permissions.py file as an initial step to centeralize
 permissions across the app.

 * Adds first and last name check to the test_func on requests to create or join a project.

 * Creates a generic wrapper decorator to be used around test_func to
 allow gradual progressive refactor of test_func towards a more
 centralized and modular permissions management solution.

 * Adds test for request creation

 Fixes #605

matthew-li commented:
Closing for now.
